### PR TITLE
Add Balancer DEX Support

### PR DESCRIPTION
**Pull request by Oozyx**

### CODE REVIEW

File: contracts/xAssetLev.sol [hunk 1]
```diff
@@ -436,12 +475,31 @@ contract xAssetLev is
             market.collateralize(depositAmount);
         }
         liquidityPool.borrow(borrowAmount);
-        // => swap usdc for baseToken
+
+        // Perform swaps (USDC -> baseToken)
         uint256 swapReturn;
```
Comment (0xlook) [inline contracts/xAssetLev.sol:480]: I don't use the usage of the `swapReturn` in `lever` function. Can we remove it?

File: contracts/xAssetLev.sol [hunk 2]
```diff
@@ -520,16 +577,37 @@ contract xAssetLev is
 
         // Swap collateral for flash loan debt
         // Max input to trade is buffer balance so that we don't trade away fees
+        uint256 flashLoanDebt = _amount + _fee;
         uint256 swapReturn;
```
Comment (0xlook) [inline contracts/xAssetLev.sol:581]: Same here

File: contracts/xAssetLev.sol [hunk 3]
```diff
@@ -754,4 +845,73 @@ contract xAssetLev is
...
+    function _swapBalancerBatchSwap(
+        IBalancerVault.SwapKind swapKind,
+        bytes32[] memory poolIds,
+        IERC20[] memory assets,
+        uint256 amount
+    ) internal returns (uint256) {
+        // build limits
+        int256[] memory limits = new int256[](assets.length);
+        for (uint256 i = 0; i < assets.length; ++i) {
+            limits[i] = type(int256).max;
```
Comment (0xlook) [inline contracts/xAssetLev.sol:887]: Maybe use `MAX_UINT` since it's used everywhere else

File: contracts/xAssetLev.sol (hunk 1 above)
Comment (Oozyx) [reply contracts/xAssetLev.sol:480]: We emit an event with `swapReturn`, we use this event when managing the xassetlev

File: contracts/xAssetLev.sol (hunk 2 above)
Comment (Oozyx) [reply contracts/xAssetLev.sol:581]: same as lever, we use `swapReturn` for the event we emit

File: contracts/xAssetLev.sol (hunk 3 above)
Comment (Oozyx) [reply contracts/xAssetLev.sol:887]: will do.

Comment (Oozyx) [reply contracts/xAssetLev.sol:887]: ah it's a bit different, limits is an array of `int256`, not `uint256`

Comment (0xlook) [reply contracts/xAssetLev.sol:887]: I see

File: contracts/xAssetLev.sol (hunk 1 above)
Comment (0xlook) [reply contracts/xAssetLev.sol:480]: Ah i see, actually the `swapReturn` has't updated in link #490

File: contracts/xAssetLev.sol [hunk 4]
```diff
@@ -520,16 +577,37 @@ contract xAssetLev is
...
+        if (exchange == Exchange.UniSingleHop) {
+            swapReturn = _swapInputForExactOutput(address(baseToken), address(usdc), getBufferBalance(), flashLoanDebt);
+        } else if (exchange == Exchange.UniMultiHop) {
+            swapReturn = _swapInputForExactOutputMultiHop(tradeData, getBufferBalance(), flashLoanDebt);
+        } else if (exchange == Exchange.BalancerSingleSwap) {
+            // decode trade data
+            bytes32 poolId = abi.decode(tradeData, (bytes32));
+
+            // make the swap
+            _swapBalancerSingleSwap(
```
Comment (0xlook) [inline contracts/xAssetLev.sol:591]: Also need to update the `swapReturn` here

File: contracts/xAssetLev.sol (hunk 1 above)
Comment (Oozyx) [reply contracts/xAssetLev.sol:480]: updated

File: contracts/xAssetLev.sol (hunk 4 above)
Comment (Oozyx) [reply contracts/xAssetLev.sol]: updated

**Review by 0xlook — APPROVED** [review]

**Pull request merged**